### PR TITLE
Llama index flavor base structure and object serialization functionality (PR 1)

### PR DESCRIPTION
**Pull request by michael-berk**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/michael-berk/mlflow/pull/12606?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12606/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12606
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
* Flavor boilerplate
* A general object serialization module
* Object testing fixtures

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

**Review by B-Step62 — COMMENTED** [review]
Thank you so much for working on this, @michael-berk!! Overall implementation looks great, left some comments but mostly minor cosmetic suggestions.

File: mlflow/llama_index/__init__.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+        return llama_index.core.__version__
+    except ImportError:
+        raise ImportError("The llama_index module is not installed.")
+    except AttributeError:
+        raise AttributeError("The llama_index module does not have a __version__ attribute.")
+
+
+@format_docstring(LOG_MODEL_PARAM_DOCS.format(package_name=FLAVOR_NAME))
+def save_model(
+    index,
```
[the remark this answers is not in the document]
Comment (michael-berk) [reply mlflow/llama_index/__init__.py:89]: Yes good callout.

I'm hoping with the recent additions of `to/from_dict()`, we'll be able to support serialization of many llama index objects. Path to model definition code (hopefully) won't be necessary for most use cases.

Either way, I'm working on PRs that handle these cases. We can rename the parameter with that PR (if that works for you)

File: mlflow/llama_index/__init__.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    settings_path = os.path.join(local_model_path, _SETTINGS_DIRECTORY)
+    # NB: Settings is a singleton and can be loaded via llama_index.core.Settings
+    deserialize_settings(settings_path)
+
+    return _load_model(local_model_path, flavor_conf)
+
+
+def _load_pyfunc(path, model_config: Optional[Dict[str, Any]] = None):
+    flavor_conf = _get_flavor_configuration(model_path=path, flavor_name=FLAVOR_NAME)
+    engine_type = model_config.pop("engine_type")
```
[the remark this answers is not in the document]
Comment (michael-berk) [reply mlflow/llama_index/__init__.py:333]: TODO

File: mlflow/llama_index/serialize_objects.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+def _object_to_dict(obj: object) -> Dict[str, any]:
+    for k, v in OBJECT_DICT_METHOD_MAP.items():
+        if isinstance(obj, k):
+            if not hasattr(obj, v[0]):
+                raise AttributeError(
+                    f"Object {obj} was inferred to be of type {k} but does not "
+                    "have a {v[0]} method. Ensure that `OBJECT_DICT_METHOD_MAP` is "
+                    "correct and the object is of type {k}."
+                )
```
[the remark this answers is not in the document]
Comment (michael-berk) [reply mlflow/llama_index/serialize_objects.py:33]: Great call! I wasn't aware of `INTERNAL_ERROR`. Makes a ton of sense

File: mlflow/llama_index/serialize_objects.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+def _sanitize_api_key(object_as_dict: Dict[str, str]) -> Dict[str, str]:
+    keys_to_remove = [k for k in object_as_dict.keys() if "api_key" in k.lower()]
+
+    for k in keys_to_remove:
+        if object_as_dict.pop(k, None):
+            _logger.info(
+                "API key removed from object serialization. At inference time,"
+                " the key must be passed as an environment variable or via inference"
+                " parameters."
```
[the remark this answers is not in the document]
Comment (michael-berk) [reply mlflow/llama_index/serialize_objects.py:64]: TODO: ensure that model config is applied to the correct location

File: mlflow/llama_index/serialize_objects.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+
+def object_to_dict(o: object) -> None:
+    try:
+        o_state_as_dict = _object_to_dict(o)
+        o_state_as_dict = _sanitize_api_key(o_state_as_dict)
+        o_state_as_dict.pop("class_name")
+    except AttributeError as e:
+        if "does not have a supported deserialization method" in str(e):
+            _logger.info(str(e))
```
[the remark this answers is not in the document]
Comment (michael-berk) [reply mlflow/llama_index/serialize_objects.py:77]: Yeah this makes sense. Depending on the index, not serializing an object can change behavior. The warnings might be verbose, but it's better to err on the side of transparency for this (IMO). Thanks!

File: tests/llama_index/_llama_index_test_fixtures.py
[the remark this answers is not in the document]
Comment (michael-berk) [reply tests/llama_index/_llama_index_test_fixtures.py:1]: 🤯

File: tests/llama_index/test_llama_index_model_export.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    ],
+)
+def test_llama_index_native_save_and_load_model(request, index_fixture, tmp_path):
+    index = request.getfixturevalue(index_fixture)
+    mlflow.llama_index.save_model(index, tmp_path, engine_type="query")
+    loaded_model = mlflow.llama_index.load_model(tmp_path)
+
+    assert type(loaded_model) == type(index)
+    assert loaded_model.as_query_engine().query("Spell llamaindex").response != ""
+
```
[the remark this answers is not in the document]
Comment (michael-berk) [reply tests/llama_index/test_llama_index_model_export.py:33]: TODO: check if this is in the next PR

File: mlflow/llama_index/serialize_objects.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (michael-berk) [reply mlflow/llama_index/serialize_objects.py:62]: Will handle in next PR

File: tests/llama_index/test_llama_index_model_export.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (michael-berk) [reply tests/llama_index/test_llama_index_model_export.py:20]: Confirmed - in the next PR. I'll also slowly be adding better coverage for integrations (datbaricks llm, openai embed model, etc.)

**Comment by michael-berk** [comment]
@mlflow-automation autoformat

**Review by B-Step62 — APPROVED** [review]
LGTM!!

**Comment by B-Step62** [comment]
Merging to unblock subsequent changes. The failed `database` test is not relevant to this PR and already resolved in master.

**Pull request merged**